[GStreamer] Fix a couple unsafe-buffer-usage warnings

philn opened the pull request:
#### e948c7ca0fa128025f1655d2461f4791b6b7d81e
<pre>
[GStreamer] Fix a couple unsafe-buffer-usage warnings
<a href="https://bugs.webkit.org/show_bug.cgi?id=283405">https://bugs.webkit.org/show_bug.cgi?id=283405</a>

Reviewed by Adrian Perez de Castro and Xabier Rodriguez-Calvar.

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::fillVideoRtpCapabilities):
* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(webKitWebSrcGetProtocols):

Canonical link: <a href="https://commits.webkit.org/286906@main">https://commits.webkit.org/286906@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d34b5af291760cd34483012a21bc475626808b76

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56499 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30380 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82045 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79581 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65647 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60703 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18704 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80531 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50673 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66490 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40984 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23987 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27069 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69184 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24324 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83453 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3294 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68945 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5000 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66458 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68209 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17040 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12214 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10310 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4791 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4810 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8245 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6569 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -463,11 +463,9 @@ void registerWebKitGStreamerElements()
-            GRefPtr<GstElementFactory> avAACDecoderFactory = adoptGRef(gst_element_factory_find(elementNames[i]));
+        std::array<ASCIILiteral, 3> elementNames { "avdec_aac"_s, "avdec_aac_fixed"_s, "avdec_aac_latm"_s };
+        for (auto& elementName : elementNames) {
+            GRefPtr<GstElementFactory> avAACDecoderFactory = adoptGRef(gst_element_factory_find(elementName.characters()));
```
It should not be needed to explicitly call `ASCIILiteral::characters()` here; the class has an implicit conversion operator to `const char*`, passing the instance directly should automatically go through it:

```suggestion
            GRefPtr<GstElementFactory> avAACDecoderFactory = adoptGRef(gst_element_factory_find(elementName));
```

aperezdc commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -492,11 +490,9 @@ void registerWebKitGStreamerElements()
-                if (auto factory = adoptGRef(gst_element_factory_find(elementNames[i])))
+            std::array<ASCIILiteral, 3> elementNames = { "dashdemux2"_s, "hlsdemux2"_s, "mssdemux2"_s };
+            for (auto& elementName : elementNames) {
+                if (auto factory = adoptGRef(gst_element_factory_find(elementName.characters())))
```
Same here:

```suggestion
                if (auto factory = adoptGRef(gst_element_factory_find(elementName)))
```

aperezdc commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -880,14 +880,12 @@ static GstURIType webKitWebSrcUriGetType(GType)
     protocols[0] = "http";
     protocols[1] = "https";
     protocols[2] = "blob";
     protocols[3] = nullptr;
```
Any reason to not use an initializer list here?

```suggestion
    static std::array<const char*, 4> protocols { "http", "https", "blob" };
```

Note that you do not need to pass `nullptr` at the end, the implementation of `std::array` takes care of using the default initializer for the array positions left after consuming the initializer list.

aperezdc reviewed (commented):
The approach of using `std::array` LGTM; I left a couple of comments, though.

aperezdc commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -1109,15 +1109,13 @@ void GStreamerRegistryScanner::fillVideoRtpCapabilities(Configuration configurat
                     auto spsAsInteger = parseInteger<uint64_t>(profileLevelId, 16).value_or(0);
-                    WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN // GLib port
-                    uint8_t sps[3];
+                    std::array<uint8_t, 3> sps;
                     sps[0] = spsAsInteger >> 16;
                     sps[1] = (spsAsInteger >> 8) & 0xff;
                     sps[2] = spsAsInteger & 0xff;
```
You may want to change this into an initializer list before landing:

```suggestion
                    std::array<uint8_t, 3> sps { spsAsInteger >> 16, (spsAsInteger >> 8) & 0xff, spsAsInteger & 0xff };
```

philn replied:
This isn't working, would need some casts maybe... I'll revert this part.

aperezdc replied:
Yeah, I wasn't sure it would work directly, that's why I had approved the PR already. I *think* assignment works because the compiler will truncate the upper bits in that case, but for initializer lists only widening is allowed. Feel free to leave the assignments if you feel that the code is clearer that way instead of adding casts around.

calvaris reviewed (approved):
What Adrián said.

philn committed:
[GStreamer] Fix a couple unsafe-buffer-usage warnings
https://bugs.webkit.org/show_bug.cgi?id=283405

Reviewed by Adrian Perez de Castro and Xabier Rodriguez-Calvar.

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::fillVideoRtpCapabilities):
* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(webKitWebSrcGetProtocols):

Canonical link: https://commits.webkit.org/286906@main
